### PR TITLE
rec: provide a few more Lua functions, most importantly record cache dump and load

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Crate additonal Lua functions to:
- Dump (some) MRU records from the record cache
- Load above dump again into the record cache
- Run some Lua on startup
- Run some Lua on graceful shutdown (`rec_control quite-nicely`)
- Create a thread running a Lua script
- Do structured logging from Lua

Provides parts of the puzzle of #8032

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [X] included documentation (including possible behaviour changes)
- [X] documented the code
- [ ] added or modified regression test(s)
- [X] added or modified unit test(s)

### CODE REVIEW

File: pdns/recursordist/docs/lua-scripting/functions.rst [hunk 1]
```diff
@@ -21,3 +21,46 @@ These are some functions that don't really have a place in one of the other cate
   Get a random number.
 
   :param int upper_bound: The upper bound. You will get a random number below this upper bound.
+
+.. function:: spawnThread(script)
+
+   .. versionadded:: 5.2.0
+
+   Spawn a thread.
```
Comment (rgacogne) [inline pdns/recursordist/docs/lua-scripting/functions.rst:29]:
```suggestion
   Spawn a new thread running the supplied Lua script in a newly created Lua context.
```

File: pdns/recursordist/docs/lua-scripting/functions.rst [hunk 2]
```diff
@@ -21,3 +21,46 @@ These are some functions that don't really have a place in one of the other cate
...
+   :param str script: The pathname of the Lua script to run.
+
+.. note::
+     The :func:`putIntoRecordCache` and :func:`getRecordCacheRecords` functions are experimental, their functionality might change in upcoming releases.
+
+.. function:: putIntoRecordCache(dump) -> int
+
+   .. versionadded:: 5.2.0
+
+   Put a record cache dump into the record cache.
```
Comment (rgacogne) [inline pdns/recursordist/docs/lua-scripting/functions.rst:40]:
```suggestion
   Load a record cache dump into the record cache.
```

File: pdns/recursordist/docs/lua-scripting/functions.rst [hunk 3]
```diff
@@ -21,3 +21,46 @@ These are some functions that don't really have a place in one of the other cate
...
+
+   :param str dump: The data in the proprietary format produced by :func:`getRecordCacheRecords`).
+   :returns: The number of records inserted into the record cache.
+
+   Some record sets might be skipped, for example when they are already present in the record cache or contain specific information not supported yet by this function.
+   If the :program:`Recursor` determines the version of the data is not compatible, it will skip loading and log an error.
```
Comment (rgacogne) [inline pdns/recursordist/docs/lua-scripting/functions.rst:46]: What is then returned? Looking at the code it's not clear to me if the exception is actually handled.

File: pdns/recursordist/recursor_cache.cc [hunk 4]
```diff
@@ -856,6 +895,375 @@ void MemRecursorCache::doPrune(time_t now, size_t keep)
...
+  message.add_bool(PBCacheEntry::required_bool_submitted, recordSet->d_submitted);
+  message.add_bool(PBCacheEntry::required_bool_tooBig, recordSet->d_tooBig);
+}
+
+size_t MemRecursorCache::getRecordSets(size_t perShard, size_t maxSize, std::string& ret)
+{
+  auto log = g_slog->withName("recordcache")->withValues("perShard", Logging::Loggable(perShard), "maxSize", Logging::Loggable(maxSize));
+  log->info(Logr::Info, "Producing cache dump");
+
+  // A size estmate is hard: size() returns the number of record *sets*. Each record set can have
```
Comment (rgacogne) [inline pdns/recursordist/recursor_cache.cc:1055]:
```suggestion
  // A size estimate is hard: size() returns the number of record *sets*. Each record set can have
```

File: pdns/recursordist/recursor_cache.cc [hunk 5]
```diff
@@ -856,6 +895,375 @@ void MemRecursorCache::doPrune(time_t now, size_t keep)
...
+{
+  address.reset();
+  protozero::pbf_message<PBComboAddress> message(reader.get_message());
+
+  // Skip all parts except address and port
+  if (message.next(PBComboAddress::required_uint32_port)) {
+    address.setPort(message.get_uint32());
+  }
+  else {
+    assert(0);
```
Comment (rgacogne) [inline pdns/recursordist/recursor_cache.cc:974]: `assert` seems a bit harsh, we could terminate the process because of a corrupted file.

**Review by rgacogne — COMMENTED** [review]
Great feature! The code looks good, but I have a few questions/comments.

File: pdns/recursordist/docs/lua-scripting/functions.rst (hunk 3 above)
Comment (omoerbeek) [reply pdns/recursordist/docs/lua-scripting/functions.rst:46]: An exception is handled at the bottom of `putRecordSets`. For the rest, see comment of latest commit 885268cac5d76cce60847cacac87380f9fbf651e

**Review by rgacogne — APPROVED** [review]

**Pull request merged**